artiq/coredevice/mirny.py

```diff
@@ -140,3 +151,143 @@ def write_ext(self, addr, length, data):
+
+        :param att_mu: attenuator setting in machine units
+        :return: attenuator setting in dB
+        """
+        return att_mu / 2
+
+    @kernel
+    def set_att_mu(self, channel, att_mu):
+        """
+        Sets attenuators on chosen shift register (channel).
+        :param channel - index of the register [0-3]
+        :param att_mu - attenuation setting in machine units [0-63]
+        """
+        if not 3 >= channel >= 0:
+            raise ValueError("channel must be 0, 1, 2 or 3")
+        if not 63 >= att_mu >= 0:
+            raise ValueError("Invalid Almazny attenuator setting")
+
+        self.att_mu[channel] = att_mu
+        self._update_register(channel)
+
+    @kernel
+    def cfg_sw(self, channel, rf_on):
+        """
+        Toggles RF switch on or off.
+        :param channel - index of the channel [0-3]
+        :param rf_on - RF output on (bool)
+        """
+        if not 3 >= channel >= 0:
+            raise ValueError("channel must be 0, 1, 2 or 3")
+        self.rf_switches[channel] = 1 if rf_on else 0
+        self._update_register(channel)
+
+    @kernel
+    def cfg_sw_all(self, rf_on):
+        """
+        Toggles all RF switches on or off.
+        :param rf_on - RF output on (bool)
+        """
+        for i in range(4):
+            self.rf_switches[i] = 1 if rf_on else 0
+        self._update_all_registers()
+
+    @kernel
+    def reg_clear(self, channel):
+        """
+        Clears content of a register.
+        :param channel - index of the channel [0-3]
+        """
+        if not 3 >= channel >= 0:
+            raise ValueError("channel must be 0, 1, 2 or 3")
+        self._send_mezz_data(0)  # reg_clear and latch are 0
+        self._send_mezz_data(ALMAZNY_REG_LATCH_BASE << channel)  # just latch is 1
+        self._send_mezz_data(ALMAZNY_REG_CLEAR)  # clear is back to 1
+
+    @kernel
+    def reg_clear_all(self):
+        """
+        Clears all registers.
+        """
+        self._send_mezz_data(0)  # reg clear and everything at 0
+        self._send_mezz_data(ALMAZNY_REG_LATCH_BASE |
+                             ALMAZNY_REG_LATCH_BASE << 1 |
+                             ALMAZNY_REG_LATCH_BASE << 2 |
+                             ALMAZNY_REG_LATCH_BASE << 3
+                            )
+        self._send_mezz_data(ALMAZNY_REG_CLEAR)
+
+    @kernel
+    def _cycle(self, data):
+        """
+        one cycle for inputting register data
+        """
+        self._send_mezz_data(ALMAZNY_REG_CLEAR | data)  # clk = 0
+        self._send_mezz_data(ALMAZNY_REG_CLEAR | ALMAZNY_SER_CLK | data)
+
+    @kernel
+    def _latch(self, ch):
+        self._send_mezz_data(ALMAZNY_REG_CLEAR | 
+                             ALMAZNY_SER_CLK |
+                             ALMAZNY_REG_LATCH_BASE << ch)
+        self._send_mezz_data(ALMAZNY_REG_CLEAR)  # reset latch
+
+    @kernel
+    def _update_register(self, ch):
+        self._send_mezz_data(ALMAZNY_REG_CLEAR)  # reset all latches
```

Comment:
Why? Aren't you loading them completely immediately afterwards? There are only 8 bits inside and you are sending 8.